### PR TITLE
Added Multi Similarity loss

### LINKED ISSUE #1017: Adding Multi-Similarity Loss to TFA losses
**Describe the feature and the current behavior/state.**

The paper's abstract:

<em>
A family of loss functions built on pair-based computation have been proposed in the literature which provide a myriad of solutions for deep metric learning. In this paper, we provide a general weighting framework for understanding recent pair-based loss functions. Our contributions are three-fold: (1) we establish a General Pair Weighting (GPW) framework, which casts the sampling problem of deep metric learning into a unified view of pair weighting
through gradient analysis, providing a powerful tool for understanding recent pair-based loss functions; (2) we show that with GPW, various existing pair-based methods can be
compared and discussed comprehensively, with clear differences and key limitations identified; (3) we propose a new loss called multi-similarity loss (MS loss) under the GPW,
which is implemented in two iterative steps (i.e., mining and weighting). This allows it to fully consider three similarities for pair weighting, providing a more principled approach for collecting and weighting informative pairs. Finally, the proposed MS loss obtains new state-of-the-art performance on four image retrieval benchmarks, where it outperforms the most recent approaches, such as ABE and HTL, by a large margin, e.g., 60.6% → 65.7% on CUB200, and 80.9% → 88.0% on In-Shop Clothes Retrieval dataset at Recall@1.  </em>

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): yes, http://openaccess.thecvf.com/content_CVPR_2019/papers/Wang_Multi-Similarity_Loss_With_General_Pair_Weighting_for_Deep_Metric_Learning_CVPR_2019_paper.pdf

- Is there already an implementation in another framework? (if so, where): yes 

In Pytorch (original):
https://github.com/MalongTech/research-ms-loss

In first reimplementation in TensorFlow (1.x):
https://github.com/geonm/tf_ms_loss

Small/quick fixes on the above code and wrap it in `keras.loss`
https://github.com/shun-lin/multi-similarity-loss-tensorflow/blob/master/Multi_Similarity_Loss_Tensorflow.ipynb

- Was it part of tf.contrib? (if so, where): No

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Loss

**Who will benefit with this feature?**
People who want to try out multi-similarity loss for their metric learning task.

### PR DESCRIPTION
**Pull request by ashutosh1919**
Fixes #1017 .  
@shun-lin, @seanpmorgan - Please review.

Also, Suggest me how to write the tests for this loss. Please suggest the way in which I should proceed to write good tests for it.

### CODE REVIEW

**Comment by ashutosh1919** [comment]
@shun-lin - I have added tests. Please review.

File: tensorflow_addons/losses/tests/multi_similarity_loss_test.py [hunk 1]
```diff
@@ -15,9 +15,143 @@
...
+    y_pred,
+    alpha=2.0,
+    beta=2.0,
+    lamb=1.0,
+    eps=0.1,
+    ms_mining=False,
+    from_logits=False,
+):
+    if from_logits:
+        y_pred = sigmoid(y_pred)
```
Comment (AakashKumarNain) [inline tensorflow_addons/losses/tests/multi_similarity_loss_test.py:44]: Use `tf.sigmoid` instead

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    y_true: TensorLike,
+    y_pred: TensorLike,
+    alpha: FloatTensorLike = 2.0,
+    beta: FloatTensorLike = 2.0,
+    lamb: FloatTensorLike = 1.0,
+    eps: FloatTensorLike = 0.1,
+    ms_mining: bool = False,
+    from_logits: bool = False,
+):
+    """Implements the multi similarity loss function.
```
Comment (shun-lin) [inline tensorflow_addons/losses/multi_similarity_loss.py:105]: NIT: I think you have copy and pasted the comments from above, may you remove the usage and shorten the description (so there isn't much duplication), also in Args you are forgetting some args.

Comment (ashutosh1919) [reply tensorflow_addons/losses/multi_similarity_loss.py:105]: done

**Comment by ashutosh1919** [comment]
@shun-lin - Done changes. Please review.

File: tensorflow_addons/losses/tests/multi_similarity_loss_test.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    neg_exp = np.exp(beta * (neg_mat - lamb))
+    neg_exp = np.where(mask_neg > 0.0, neg_exp, np.zeros_like(neg_exp))
+    pos_term = np.log(1.0 + np.sum(pos_exp, axis=1)) / alpha
+    neg_term = np.log(1.0 + np.sum(neg_exp, axis=1)) / beta
+    loss = np.mean(pos_term + neg_term)
+    return loss
+
+
+def get_labels_and_preds(out_shape=(2, 3)):
+    y_true = np.array(np.random.uniform(size=(2, 3)), dtype=np.float32)
```
Comment (autoih) [inline tensorflow_addons/losses/tests/multi_similarity_loss_test.py:77]:
```suggestion
    y_true = np.array(np.random.uniform(out_shape=(2, 3)), dtype=np.float32)
```

Comment (autoih) [inline tensorflow_addons/losses/tests/multi_similarity_loss_test.py:77]:
```suggestion
    y_true = np.array(np.random.uniform(size=out_shape), dtype=np.float32)
```

**Comment by ashutosh1919** [comment]
@autoih , Thanks for noticing mistake.

**Comment by ashutosh1919** [comment]
@shun-lin - can you manually approve CLA since there is no reply from @autoih .

**Comment by autoih** [comment]
@googlebot I fixed it.

**Comment by ashutosh1919** [comment]
@autoih , Please sign the CLA with all the email IDs you are logged into your system. I had the same issue. But it resolves the issue.

**Comment by autoih** [comment]
@googlebot I fixed it.

**Comment by autoih** [comment]
> @autoih , Please sign the CLA with all the email IDs you are logged into your system. I had the same issue. But it resolves the issue.

I think I did, but not sure why it has the issue still.

**Comment by seanpmorgan** [comment]
Apologies, I haven't had the time to review as much as I would like... but just wanted to comment that the CLA is not an issue. We can override it during merge since you're both active contributors that have signed the CLA.

**Comment by ashutosh1919** [comment]
Thanks @seanpmorgan for the comment.

**Comment by shun-lin** [comment]
@ashutosh1919 do you also want to add a tutorial (notebook) for this loss as well?

**Comment by ashutosh1919** [comment]
@shun-lin , Okay let me add.

**Comment by ashutosh1919** [comment]
Please ignore below images. Just uploaded for retrieving links to paste in notebook.

**Comment by ashutosh1919** [comment]
ms_loss_steps

ms_loss_equation

**Comment by ashutosh1919** [comment]
@shun-lin - Please review.

**Comment by ashutosh1919** [comment]
@shun-lin - Any update on this PR? Thanks!

**Comment by MathisFederico** [comment]
I would like very much to try this metric ! Is there a reason why the PR is frozen ?

**Comment by shun-lin** [comment]
Hi sorry, somehow notification for this PR fall through my gmail filter, will review! Sorry for the delay.

File: docs/tutorials/losses_multi_similarity.ipynb [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (shun-lin) [inline docs/tutorials/losses_multi_similarity.ipynb:88]: May you review this block from the notebook?

```
try:
  %tensorflow_version 2.x
except Exception:
  pass
```

is no longer necessary as colab is defaulted tf version 2.x now

File: docs/tutorials/losses_multi_similarity.ipynb [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    "# Build your input pipelines\n",
+    "train_dataset = train_dataset.shuffle(1024).batch(32)\n",
+    "train_dataset = train_dataset.map(_normalize_img)\n",
+    "\n",
+    "test_dataset = test_dataset.batch(32)\n",
+    "test_dataset = test_dataset.map(_normalize_img)"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (shun-lin) [inline docs/tutorials/losses_multi_similarity.ipynb:147]: may you remove this block as well?

```
c = True
for v in train_dataset.as_numpy_iterator():
    if c:
        print(v[1].shape)
        c = False
```

This block seems to just be testing.

File: docs/tutorials/losses_multi_similarity.ipynb (hunk 4 above)
Comment (ashutosh1919) [reply docs/tutorials/losses_multi_similarity.ipynb:88]: done

File: docs/tutorials/losses_multi_similarity.ipynb (hunk 5 above)
Comment (ashutosh1919) [reply docs/tutorials/losses_multi_similarity.ipynb]: done

**Comment by ashutosh1919** [comment]
@shun-lin - Changes made. Please review.

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+@tf.function
+def multi_similarity_loss(
+    y_true: TensorLike,
+    y_pred: TensorLike,
+    alpha: FloatTensorLike = 2.0,
+    beta: FloatTensorLike = 2.0,
+    lamb: FloatTensorLike = 1.0,
+    eps: FloatTensorLike = 0.1,
```
Comment (WindQAQ) [inline tensorflow_addons/losses/multi_similarity_loss.py:101]: our public API all use `epsilon` (core TF as well), please change it to `epsilon` for consistency. Also, `0.1` is not a very **small positive number**. Is there any reason we set it to such large value?

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      Loss float `Tensor`.
+    """
+
+    @typechecked
+    def __init__(
+        self,
+        alpha: FloatTensorLike = 2.0,
+        beta: FloatTensorLike = 50.0,
+        lamb: FloatTensorLike = 1.0,
+        eps: FloatTensorLike = 0.1,
```
Comment (WindQAQ) [inline tensorflow_addons/losses/multi_similarity_loss.py:67]: ditto

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        return multi_similarity_loss(
+            y_true,
+            y_pred,
+            self.alpha,
+            self.beta,
+            self.lamb,
+            self.eps,
+            self.ms_mining,
+            self.from_logits,
+        )
```
Comment (WindQAQ) [inline tensorflow_addons/losses/multi_similarity_loss.py:90]: do not forget to override `get_config` here.

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        mask_neg = tf.where(neg_mat > min_val - eps, mask_neg, tf.zeros_like(mask_neg))
+
+    pos_exp = tf.exp(-alpha * (pos_mat - lamb))
+    pos_exp = tf.where(mask_pos > 0.0, pos_exp, tf.zeros_like(pos_exp))
+
+    neg_exp = tf.exp(beta * (neg_mat - lamb))
+    neg_exp = tf.where(mask_neg > 0.0, neg_exp, tf.zeros_like(neg_exp))
+
+    pos_term = tf.math.log(1.0 + tf.reduce_sum(pos_exp, axis=1)) / alpha
+    neg_term = tf.math.log(1.0 + tf.reduce_sum(neg_exp, axis=1)) / beta
```
Comment (WindQAQ) [inline tensorflow_addons/losses/multi_similarity_loss.py:166]:
```suggestion
    pos_term = tf.math.log1p(tf.reduce_sum(pos_exp, axis=1)) / alpha
    neg_term = tf.math.log1p(tf.reduce_sum(neg_exp, axis=1)) / beta
```

File: tensorflow_addons/losses/multi_similarity_loss.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+# ==============================================================================
+"""Implements Multi Similarity loss."""
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import FloatTensorLike, TensorLike
+from typeguard import typechecked
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiSimilarityLoss(tf.keras.losses.Loss):
```
Comment (WindQAQ) [inline tensorflow_addons/losses/multi_similarity_loss.py:23]: Any reason that we do not subclass `LossFunctionWrapper`?

Comment (ashutosh1919) [reply tensorflow_addons/losses/multi_similarity_loss.py:23]: Python private APIs not allowed.

Comment (WindQAQ) [reply tensorflow_addons/losses/multi_similarity_loss.py:23]: https://github.com/tensorflow/addons/blob/40d8e59cbff54204affbfc44ae3dbf8021aab427/tensorflow_addons/utils/keras_utils.py#L24

Comment (ashutosh1919) [reply tensorflow_addons/losses/multi_similarity_loss.py:23]: Got it. Will transform the file. Give me sometime.

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
